File: apps/ruby/lib/getting_started_ml/chapter15/validation.rb
# frozen_string_literal: true

require "json"

module GettingStartedMl
  module Chapter15
    # プレゼンテーション層の入口。要求の JSON を読み、検証してドメインの値にする。
    #
    # JSON.parse は型を決めずに Hash・Integer・Float・String を返すので、Rust 版の serde が
    # 読み込みと同時にしていた型の確認を、列ごとの型の表で自分で行う。
    module Validation
      # 検証の結果。理由の一覧が空なら正しく、value にドメインの値が入る。
      Validated = Data.define(:value, :errors) do
        def valid?
          errors.empty?
        end
      end

      # 興行収入の予測の要求の列と型。原作の有無は整数で受け取る。
      MOVIE_TYPES = { "sns1" => Numeric, "sns2" => Numeric, "actor" => Numeric, "original" => Integer }.freeze

      # 生存の予測の要求の列と型。
      PASSENGER_TYPES = { "pclass" => Integer, "sex" => String, "age" => Numeric, "sib_sp" => Integer,
                          "parch" => Integer, "fare" => Numeric, "embarked" => String }.freeze

      module_function

      # 本文を JSON のオブジェクトとして読み、列の型を確かめる。読めないか型が合わなければ nil を返す。
      # null は「値が無い」とみなし、表に無い列は無視する。
      def read_json(body, types)
        request = JSON.parse(body)
        return nil unless request.is_a?(Hash)

        request.all? { |field, value| typed?(types[field], value) } ? request : nil
      rescue JSON::ParserError
        nil
      end

      # 値が型に合うかどうか。null と、表に無い列（型が nil）は問わない。
      def typed?(type, value)
        type.nil? || value.nil? || value.is_a?(type)
      end

      # 理由（nil は問題なし）を集め、1 つも無ければブロックで値を作る。
      def validate(reasons)
        errors = reasons.compact

        Validated.new(value: errors.empty? ? yield : nil, errors:)
      end

      # 値が無ければ理由を返す。
      def required(request, field)
        request[field].nil? ? "#{field} は必須です" : nil
      end

      # 値が負であれば理由を返す。値が無ければ何も言わない（required の担当）。
      def not_negative(request, field)
        value = request[field]
        !value.nil? && value.negative? ? "#{field} は 0 以上にしてください" : nil
      end

      # 値が選択肢に無ければ理由を返す。値が無ければ何も言わない。
      def one_of(request, field, allowed)
        value = request[field]
        value.nil? || allowed.include?(value) ? nil : "#{field} は #{allowed.join('、')} のどれかにしてください"
      end

      # 検証して、正しければ映画の特徴量にする。
      def movie(request)
        fields = %w[sns1 sns2 actor original]
        reasons = fields.map { |field| required(request, field) } +
                  %w[sns1 sns2 actor].map { |field| not_negative(request, field) } +
                  [one_of(request, "original", [0, 1])]

        validate(reasons) { Movie.new(**fields.to_h { |field| [field.to_sym, request[field]] }) }
      end

      # 検証して、正しければ乗客の特徴量にする。省略された列は空文字列（欠損値）にする。
      def passenger(request)
        validate(passenger_reasons(request)) do
          Passenger.new(**PASSENGER_TYPES.keys.to_h { |field| [field.to_sym, request[field].to_s] })
        end
      end

      # 乗客の要求の理由。年齢と乗船港は省略できる。
      def passenger_reasons(request)
        %w[pclass sex sib_sp parch fare].map { |field| required(request, field) } +
          [one_of(request, "pclass", [1, 2, 3]), one_of(request, "sex", %w[female male])] +
          %w[age sib_sp parch fare].map { |field| not_negative(request, field) } +
          [one_of(request, "embarked", %w[C Q S])]
      end
    end
  end
end

File: apps/elixir/lib/getting_started_ml/chapter15/validation.ex
defmodule GettingStartedMl.Chapter15.Validation do
  @moduledoc """
  第 15 章のプレゼンテーション層。要求の JSON を読み、検証してドメインの値にする。

  Jason は JSON の値をそのまま Elixir のマップ・数値・文字列にするので、
  Java 版のように「読み込みと同時に型が確かめられる」ことはない。型の確認は自分で書く。
  """

  @movie_types %{"sns1" => :number, "sns2" => :number, "actor" => :number, "original" => :integer}

  @passenger_types %{
    "pclass" => :integer,
    "sex" => :string,
    "age" => :number,
    "sib_sp" => :integer,
    "parch" => :integer,
    "fare" => :number,
    "embarked" => :string
  }

  @original_values [0, 1]
  @passenger_classes [1, 2, 3]
  @sexes ["female", "male"]
  @ports ["C", "Q", "S"]

  @doc "興行収入の予測の要求の列と型。原作の有無は整数で受け取る（1.5 を弾くため）。"
  def movie_types, do: @movie_types

  @doc "生存の予測の要求の列と型。年齢と乗船港は省略できる。"
  def passenger_types, do: @passenger_types

  @doc """
  本文を JSON のオブジェクトとして読み、列の型を確かめる。

  読めれば `{:ok, マップ}`、読めないか型が合わなければ `:error` を返す。
  """
  def read_json(body, types) do
    with {:ok, request} <- Jason.decode(body),
         true <- is_map(request),
         true <- Enum.all?(request, fn {field, value} -> typed?(Map.get(types, field), value) end) do
      {:ok, request}
    else
      _other -> :error
    end
  end

  # 値が型に合うかどうか。null と、表に無い列（型が nil）は問わない。
  defp typed?(_type, nil), do: true
  defp typed?(nil, _value), do: true
  defp typed?(:number, value), do: is_number(value)
  defp typed?(:integer, value), do: is_integer(value)
  defp typed?(:string, value), do: is_binary(value)

  ## 検証の規則。問題が無ければ nil を、あれば理由を返す

  @doc "必須の列が空なら理由を返す。"
  def required(field, nil), do: "#{field} は必須です"
  def required(_field, _value), do: nil

  @doc "負の数なら理由を返す。"
  def not_negative(field, value) when is_number(value) and value < 0,
    do: "#{field} は 0 以上にしてください"

  def not_negative(_field, _value), do: nil

  @doc "選択肢の外の値なら理由を返す。選択肢は並べ替えて表示する。"
  def one_of(_field, nil, _allowed), do: nil

  def one_of(field, value, allowed) do
    if value in allowed do
      nil
    else
      "#{field} は #{Enum.map_join(Enum.sort(allowed), "、", &to_string/1)} のどれかにしてください"
    end
  end

  @doc """
  理由（nil は問題なし）を集め、1 つも無ければ `build` で値を作る。

  Java 版の `sealed interface Validated` と違い、値と理由を両方持つマップで表す。
  `build` を関数で受け取るのは、理由があるときに値を作らせないためである。
  """
  def validate(reasons, build) do
    case Enum.reject(reasons, &is_nil/1) do
      [] -> %{value: build.(), errors: []}
      errors -> %{value: nil, errors: errors}
    end
  end

  @doc "検証の結果が正しいかどうか。"
  def valid?(%{errors: errors}), do: errors == []

  @doc "検証して、正しければ映画の特徴量にする。"
  def movie(request) do
    %{"sns1" => sns1, "sns2" => sns2, "actor" => actor, "original" => original} =
      Map.merge(%{"sns1" => nil, "sns2" => nil, "actor" => nil, "original" => nil}, request)

    validate(
      [
        required("sns1", sns1),
        required("sns2", sns2),
        required("actor", actor),
        required("original", original),
        not_negative("sns1", sns1),
        not_negative("sns2", sns2),
        not_negative("actor", actor),
        one_of("original", original, @original_values)
      ],
      fn ->
        %{
          sns1: sns1 / 1,
          sns2: sns2 / 1,
          actor: actor / 1,
          original: original
        }
      end
    )
  end

  @doc "検証して、正しければ乗客の特徴量にする。年齢と乗船港は省略できる。"
  def passenger(request) do
    %{
      "pclass" => pclass,
      "sex" => sex,
      "age" => age,
      "sib_sp" => sib_sp,
      "parch" => parch,
      "fare" => fare,
      "embarked" => embarked
    } = Map.merge(Map.new(Map.keys(@passenger_types), &{&1, nil}), request)

    validate(
      [
        required("pclass", pclass),
        required("sex", sex),
        required("sib_sp", sib_sp),
        required("parch", parch),
        required("fare", fare),
        one_of("pclass", pclass, @passenger_classes),
        one_of("sex", sex, @sexes),
        not_negative("age", age),
        not_negative("sib_sp", sib_sp),
        not_negative("parch", parch),
        not_negative("fare", fare),
        one_of("embarked", embarked, @ports)
      ],
      fn ->
        %{
          pclass: pclass,
          sex: sex,
          age: age && age / 1,
          sib_sp: sib_sp,
          parch: parch,
          fare: fare / 1,
          embarked: embarked
        }
      end
    )
  end
end
